### PR TITLE
Adds rvalue overloads for var matrices

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

This adds all the `rvalue()` overloads needed for `var<matrix>` and relaxes the templates on the current overloads that can handle `var<matrix>`

The only rvalue overloads needed are the for `index_multi` since we have to have a callback to propogate adjoints.

#### Intended Effect

Allow subsetting in Stan models when using `var<matrix>`

#### How to Verify

Manual tests have been added for all single index and all indices on the right hand side for matrices. Then a larger automated test is run that tests all pair combinations of indices.

#### Documentation

Docs added for each new overload and original docs for rvalue overloads updated for when they can accept either Eigen matrices or `var<matrix>` types

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Steve Bronder

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/stan/model/indexing/rvalue_varmat.hpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+  for (int j = 0; j < ret_cols; ++j) {
+    check_range("matrix[..., multi] col index", name, x.cols(), col_idx[j]);
+    --col_idx[j];
+    x_ret_val.col(j) = rvalue(x.val().col(col_idx[j]), index_list(idxs.head_),
+                              name, depth + 1);
+  }
+  var_value<Eigen::Matrix<double, -1, -1>> x_ret(x_ret_val);
+  // index_multi is the only index with dynamic memory so head is safe
+  reverse_pass_callback(
+      [head = idxs.head_, x, x_ret, col_idx, name, depth]() mutable {
```
Comment (t4c1) [inline src/stan/model/indexing/rvalue_varmat.hpp:308]: I dont see name or depth used in the lambda.

**Review by t4c1 — CHANGES_REQUESTED** [review]

File: src/stan/model/indexing/rvalue_varmat.hpp [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+  reverse_pass_callback([x, x_ret, row_idx]() mutable {
+    for (Eigen::Index i = 0; i < row_idx.size(); ++i) {
+      x.adj().coeffRef(row_idx[i]) += x_ret.adj().coeffRef(i);
+    }
+  });
+  return x_ret;
+}
+
+/**
+ * Return the specified matrix at the specified multi index.
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply src/stan/model/indexing/rvalue_varmat.hpp:85]: Yes i also went through and fixed up the docs for `idx` so it was more obvious what each one did

File: src/stan/model/indexing/rvalue_varmat.hpp [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+  using stan::math::var_value;
+  using arena_std_vec = std::vector<int, arena_allocator<int>>;
+  const Eigen::Index x_size = x.size();
+  const auto ret_size = idxs.head_.ns_.size();
+  arena_t<value_type_t<Vec>> x_ret_vals(ret_size);
+  arena_std_vec row_idx(idxs.head_.ns_.begin(), idxs.head_.ns_.end());
+  for (int i = 0; i < ret_size; ++i) {
+    check_range("vector[multi] assign range", name, x_size, row_idx[i]);
+    --row_idx[i];
+    x_ret_vals.coeffRef(i) = x.vi_->val_.coeffRef(row_idx[i]);
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply src/stan/model/indexing/rvalue_varmat.hpp:73]: Went through and checked this over both

**Review by t4c1 — APPROVED** [review]

**Comment by rok-cesnovar** [comment]
@SteveBronder cmdstan Windows Rtools 3.5 test is failing after this PR was merged. See https://github.com/stan-dev/cmdstan/runs/1591785707 and https://pipelines.actions.githubusercontent.com/IWD9CSoapEZ4VsVQhL3f8rgS1MWkcN50p1yldEctrAl8hmLKx1/_apis/pipelines/1/runs/286/signedlogcontent/3?urlExpires=2020-12-22T10%3A24%3A24.1063345Z&urlSigningMethod=HMACV1&urlSignature=u1xcmK7mG%2B9qOtorsydy4sWoOcopJ2cvK9duGu7NorU%3D

```cpp
--- Translating Stan model to C++ code ---
bin/stanc.exe  --o=examples/bernoulli/bernoulli.hpp examples/bernoulli/bernoulli.stan

--- Compiling, linking C++ code ---
g++ -std=c++1y -m64 -D_REENTRANT -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs -Wno-int-in-bool-context -Wno-attributes -Wno-ignored-attributes      -I stan/lib/stan_math/lib/tbb_2019_U8/include   -O3 -I src -I stan/src -I lib/rapidjson_1.1.0/ -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.9 -I stan/lib/stan_math/lib/boost_1.72.0 -I stan/lib/stan_math/lib/sundials_5.5.0/include  -D_USE_MATH_DEFINES  -DBOOST_DISABLE_ASSERTS         -c  -x c++ -o examples/bernoulli/bernoulli.o examples/bernoulli/bernoulli.hpp
In file included from stan/src/stan/model/indexing/lvalue.hpp:8:0,
                 from stan/src/stan/model/indexing.hpp:7,
                 from stan/src/stan/model/model_header.hpp:15,
                 from examples/bernoulli/bernoulli.hpp:3:
stan/src/stan/model/indexing/rvalue_at.hpp: In function 'constexpr int stan::model::rvalue_at(int, const stan::model::index_min_max&)':
stan/src/stan/model/indexing/rvalue_at.hpp:75:1: error: body of constexpr function 'constexpr int stan::model::rvalue_at(int, const stan::model::index_min_max&)' not a return-statement
 }
 ^
```
The error message is not very helpful unfortunately.

This is not super urgent as develop tests will not fail due to this, but would still be great to address soon.

**Comment by SteveBronder** [comment]
Oh yes I know how to fix this I'll put up a PR right now. I think it's because windows does not have full C++14 support so it's failing when trying to eval this

**Comment by SteveBronder** [comment]
@rok-cesnovar just put this up, will fix the above

https://github.com/stan-dev/stan/pull/2993

**Pull request merged**